### ISSUE TITLE
Extracting LDL decomposition from sampling solver

### ISSUE THREAD

**Issue by jmblpati** [danspielman/Laplacians.jl#3]
I am working on my own implementation of the Laplacian system solver from http://arxiv.org/pdf/1605.02353v1.pdf . I have run into issues, and so I am looking at the reference implementation provided here for guidance. However, I have some questions about the way that the implementation here produces the matrices U and D. The procedure samplingLapSolver appears to compute the matrices U and D as in the paper and  do CG with f(x) = U \ (D \ (U' \x )) as a preconditioner. However, when the matrices U and D are multiplied as U'_D_U, the results do not match the original matrix within the factor provided in the paper:

`defaultSamplingParams = samplingParams(0.5, 0.02, 1e0, 1000, 20,
                                false,false,1e-3)`
`#even if beta is changed back to 1e3, the eigenvalue bounds do not hold (in fact, they are even worse)`
`U = Laplacians.buildSolver(grid3(4))[3]`
`d = Laplacians.buildSolver(grid3(4))[4];`
`w = Laplacians.lap(grid3(4));`
`z1 = U'*diagm(d)*U`
`test1 = sort(eig(full(3*w/2 - z1))[1])[1]`
`test2 = sort(eig(full(z1- w/2))[1])[1]`
`println(test1, " ", test2) #these need to both be positive`
`-6.353282856003461 -2.137250265528108`

Multiplying by the permutation matrix doesn't change this
`ord = Laplacians.buildSolver(grid3(4))[5];`
`for i = 1:64`
`ordmat[i,ord[i]] = 1` 
`end`
`z1 = ordmat*U'*diagm(d)*U*ordmat'`
`compute tests`
-5.893602233549341 -2.6099184213847844

grid3 is just an alias for wGrid3, taken with fixed unit edge weight.

Can someone please explain how to extract the explicit cholesky decomposition from the solver?

**Comment by serbanstan** [comment]
The way you described is the correct way to generate the U and D matrices. The problem lies in how you called the solver.

You get the U and D matrices with different calls to the solver. As no random seed is set inside the solver, these matrices can correspond to different decompositions. A solution is to just get them with the same call. 

Also, on my machine, overwriting defaultSamplingParams doesn't work. A workaround is to just pass the parameters to the solver when you call it. 

Finally, the solver gives better approximations the bigger the sampConst parameter gets (the second parameter in samplingParams). Try increasing that if you have further issues.

Here's an example of implementing the above changes into your code.

```
using Laplacians
params = samplingParams(0.5, 0.2, 1e0, 1000, 20, false,false,1e-3)
A = wGrid3(4);
F,gOp,U,d,ord,_,_ = Laplacians.buildSolver(A,params=params);
z1 = U'*diagm(d)*U;
w = lap(A);
test1 = sort(eig(full(3*w/2 - z1))[1])[1]
test2 = sort(eig(full(z1- w/2))[1])[1];
println(test1, " ", test2)
-3.0208822381589635 -1.7160808074269742
```

But, if you use the permutation matrix

```
ordmat = zeros(A.n,A.m);
for i = 1:A.n
    ordmat[i,ord[i]] = 1 
end
z1 = ordmat'*U'*diagm(d)*U*ordmat;
test1 = sort(eig(full(3*w/2 - z1))[1])[1]
test2 = sort(eig(full(z1- w/2))[1])[1];
println(test1, " ", test2)
1.3245769913509612e-15 4.666210924884499e-16
```

**Comment by jmblpati** [comment]
Ok, that makes sense. I guess I missed the ordering permutation on my first reading of the code.

However, I have a question about ord. The way that it is used seems to parallel the random permutation pi in the setup of the paper. However, if ord is instead set to an actual random permutation, the approximation guarantee no longer holds. This behavior is seen even if beta is set to 1 and the low stretch tree is eliminated (instead of taking the smaller of the pairwise resistance and rho, we always take rho for that weight). Do you have any explanation for this?

**Comment by serbanstan** [comment]
I think the bounds hold even with a random permutation and no tree. Maybe try increasing sampConst to 2.0 from 0.2 in the above code. It might be the case that the inverse DFS order is particularly good for the type of algorithm we are using.

**Comment by jmblpati** [comment]
Yeah, I increased the sampling constant from 0.2 to 0.4 and that fixed it! I am now getting the results that I was expecting. Thanks a lot. 

I have modified the sampling solver from Laplacians.jl to make it closer to the original paper and a bit easier to understand (if a bit slower as well). https://github.com/jmblpati/Laplacians.jl/blob/master/src/samplingSolver.jl I just thought I'd leave this here in case it might be helpful to someone in the future. 

Thanks for all your help!
